core/kernel/tee_ta_manager.c

```diff
@@ -923,6 +918,50 @@ static void tee_ta_update_session_utime(bool suspend)
 		sbuf->usr_entered = now;
 	}
 }
+#endif
+
+#if defined(CFG_TA_FTRACE_SUPPORT)
+static void ftrace_update_session_utime(bool suspend, struct tee_ta_session *s,
+					uint64_t now)
+{
+	struct ftrace_buf *fbuf = NULL;
+	uint32_t i = 0;
+
+	fbuf = s->fbuf;
+	if (!fbuf)
+		return;
+
+	if (suspend) {
+		fbuf->suspend_time = now;
+	} else {
+		for (i = 0; i <= fbuf->ret_idx; i++)
+			fbuf->begin_time[i] += now - fbuf->suspend_time;
+	}
+}
+#endif
+
+/*
+ * Update user-mode CPU time for the current session
+ * @suspend: true if session is being suspended (leaving user mode), false if
+ * it is resumed (entering user mode)
+ */
+static void tee_ta_update_session_utime(bool suspend)
+{
+	struct tee_ta_session *s = NULL;
+	uint64_t now = 0;
+
+	if (tee_ta_get_current_session(&s) != TEE_SUCCESS)
+		return;
+
+	now = read_cntpct();
+
+#if defined(CFG_TA_GPROF_SUPPORT)
+	gprof_update_session_utime(suspend, s, now);
+#endif
+#if defined(CFG_TA_FTRACE_SUPPORT)
+	ftrace_update_session_utime(suspend, s, now);
+#endif
```

Comment:
minor: you could easily avoid those `#ifdefs` with static inline funcs in the header file (as we usually do).

Reply:
In this case the functions to be called under `#ifdefs` are in the same `tee_ta_manager.c` file. So there isn't any declaration in the header. One another way could be to provide empty definitions in `#else` part in `.c` file, your views?

Reply:
Ah, true. Please don't bother. Not worth it.